packages/vaadin-virtual-list/src/virtualizer-iron-list-adapter.js

```diff
@@ -0,0 +1,430 @@
+    if (fviOffsetBefore !== undefined && fviOffsetAfter !== undefined) {
+      this._scrollTop += fviOffsetBefore - fviOffsetAfter;
+    }
+
+    if (!this.elementsContainer.children.length) {
+      requestAnimationFrame(() => this._resizeHandler());
+    }
+  }
+
+  get size() {
+    return this.__size;
+  }
+
+  /** @private */
+  get _scrollTop() {
+    return this.scrollTarget.scrollTop;
+  }
+
+  /** @private */
+  set _scrollTop(top) {
+    this.scrollTarget.scrollTop = top;
+  }
+
+  /** @private */
+  get items() {
+    return {
+      length: Math.min(this.size, MAX_VIRTUAL_COUNT)
+    };
+  }
+
+  /** @private */
+  get offsetHeight() {
+    return this.scrollTarget.offsetHeight;
+  }
+
+  /** @private */
+  get $() {
+    return {
+      items: this.scrollContainer
+    };
+  }
+
+  /** @private */
+  updateViewportBoundaries() {
+    const styles = window.getComputedStyle(this.scrollTarget);
+    this._scrollerPaddingTop = this.scrollTarget === this ? 0 : parseInt(styles['padding-top'], 10);
+    this._isRTL = Boolean(styles.direction === 'rtl');
+    this._viewportWidth = this.elementsContainer.offsetWidth;
+    this._viewportHeight = this.scrollTarget.offsetHeight;
+    this._scrollPageHeight = this._viewportHeight - this._scrollLineHeight;
+    this.grid && this._updateGridMetrics();
+  }
+
+  /** @private */
+  setAttribute() {}
+
+  /** @private */
+  _createPool(size) {
+    const physicalItems = this.createElements(size);
+    const fragment = document.createDocumentFragment();
+    physicalItems.forEach((el) => {
+      el.style.position = 'absolute';
+      fragment.appendChild(el);
+      this.__resizeObserver.observe(el);
+    });
+    this.elementsContainer.appendChild(fragment);
+    return physicalItems;
+  }
+
+  /** @private */
+  _assignModels(itemSet) {
+    this._iterateItems((pidx, vidx) => {
+      const el = this._physicalItems[pidx];
+      el.hidden = vidx >= this.size;
+      if (!el.hidden) {
+        el.__virtualIndex = vidx + (this._vidxOffset || 0);
+        this.updateElement(el, el.__virtualIndex);
+      }
+    }, itemSet);
+  }
+
+  /** @private */
+  _isClientFull() {
+    // Workaround an issue in iron-list that can cause it to freeze on fast scroll
+    setTimeout(() => (this.__clientFull = true));
+    return this.__clientFull || super._isClientFull();
+  }
+
+  /** @private */
+  translate3d(_x, y, _z, el) {
+    el.style.transform = `translateY(${y})`;
+  }
+
+  /** @private */
+  toggleScrollListener() {}
+
+  _scrollHandler() {
+    this._adjustVirtualIndexOffset(this._scrollTop - (this.__previousScrollTop || 0));
+
+    super._scrollHandler();
+
+    if (this.reorderElements) {
+      this.__scrollReorderDebouncer = Debouncer.debounce(
+        this.__scrollReorderDebouncer,
+        timeOut.after(this.timeouts.SCROLL_REORDER),
+        () => this.__reorderElements()
+      );
+    }
+
+    this.__previousScrollTop = this._scrollTop;
+  }
+
+  /** @private */
+  __onWheel(e) {
```

Comment:
This and the following few functions were moved from [here](https://github.com/vaadin/web-components/blob/b163d3a8c200e6a9064e03b7ca1c4a34ebb4f3d9/packages/vaadin-grid/src/vaadin-grid-scroll-mixin.js#L133). Their purpose is to support different wheel event delta modes (LINE, PAGE) and also to make wheel scrolling UX nicer; once you reach the end of the scrollable container, its scrollable host container doesn't immediately continue scrolling.